### PR TITLE
feat(minifier): finish implementing folding array expressions

### PR DESCRIPTION
**Pull request by camc314**

### CODE REVIEW

**Comment by camc314** [comment]
* **#6575** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6575?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @camc314 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: crates/oxc_minifier/src/ast_passes/peephole_remove_dead_code.rs [hunk 1]
```diff
@@ -379,17 +443,16 @@ mod test {
...
 
     #[test]
-    #[ignore]
     fn test_array_literal_containing_spread() {
         fold_same("([...c])");
         fold("([4, ...c, a])", "([...c])");
-        fold("([foo(), ...c, bar()])", "(foo(), [...c], bar())");
-        fold("([...a, b, ...c])", "([...a], [...c])");
+        //fold("([foo(), ...c, bar()])", "(foo(), [...c], bar())");
+        //fold("([...a, b, ...c])", "([...a], [...c])");
```
Comment (camc314) [inline crates/oxc_minifier/src/ast_passes/peephole_remove_dead_code.rs:455]: my core currently outputs `foo(), bar(), [...c];` for the first commented test
and `[...a, ...c];` for the second

i think this is ok to just change? thoughts?

File: crates/oxc_minifier/src/ast_passes/peephole_remove_dead_code.rs [hunk 2]
```diff
@@ -379,17 +443,16 @@ mod test {
     fn test_array_literal() {
         fold("([])", "");
         fold("([1])", "");
-        // fold("([a])", "a");
-        // fold("([foo()])", "foo()");
+        fold("([a])", "a");
```
Comment (camc314) [inline crates/oxc_minifier/src/ast_passes/peephole_remove_dead_code.rs:446]: Ci currently fails on this one - i remove the `a` but closure keeps it
closures behaviour doesn't make sense to me here - on L455, closure removes `b` ?

Comment (camc314) [reply crates/oxc_minifier/src/ast_passes/peephole_remove_dead_code.rs]: ignore this - looks like the tests were ported over incorrectly:

https://github.com/google/closure-compiler/blob/f888b14e237f6f18f9cb66eeef3e79f398c45a68/test/com/google/javascript/jscomp/PeepholeRemoveDeadCodeTest.java#L1477-L1479

File: crates/oxc_minifier/src/ast_passes/peephole_remove_dead_code.rs (hunk 1 above)
Comment (Boshen) [reply crates/oxc_minifier/src/ast_passes/peephole_remove_dead_code.rs:455]: There may be different semantics? You need to play safe 😅 

What does other minifiers compress to?

Comment (camc314) [reply crates/oxc_minifier/src/ast_passes/peephole_remove_dead_code.rs:455]: ah yeah you're right.
spreading `c` could mutate bar

i think the our fix for the 2nd case is fine.


| input               | esbuild        | swc            | terser   | closure             |
| ------------------- | -------------- | -------------- | -------- | ------------------- |
| `([...a, b, ...c])` | `[...a,...c];` | `[...a,...c];` | `a,b,c;` | `([...a], [...c])"` |

esbuild/swc have the same behaviour. i think terser's behaviour is just wrong.

**Pull request merged**